### PR TITLE
(core) upgrade lodash to 4.x

### PR DESCRIPTION
**Pull request by icfantv**
This is a massive change.  I've looked at https://github.com/lodash/lodash/wiki/Changelog and believe I've addressed the compatibility issues but won't even remotely swear to it being flawless.

Migration strategy was to look at methods that were removed, renamed, and whose functionality has been split out.

### CODE REVIEW

**Comment by icfantv** [comment]
@spinnaker/google-reviewers @zanthrash @anotherchrisberry PTAL and let me know when (not if) you have questions.

File: app/scripts/modules/azure/instance/azureInstanceType.service.js [hunk 1]
```diff
@@ -225,23 +226,23 @@ module.exports = angular.module('spinnaker.azure.instanceType.service', [
...
 
-          stats.cpu.min = Math.min(stats.cpu.min, cpuMin);
-          stats.cpu.max = Math.max(stats.cpu.max, cpuMax);
-          stats.memory.min = Math.min(stats.memory.min, memoryMin);
-          stats.memory.max = Math.max(stats.memory.max, memoryMax);
-          stats.storage.min = Math.min(stats.storage.min, storageMin);
-          stats.storage.max = Math.max(stats.storage.max, storageMax);
-          stats.costFactor.min = Math.min(stats.costFactor.min, costFactorMin);
-          stats.costFactor.max = Math.max(stats.costFactor.max, costFactorMax);
+          stats.cpu.min = Math.minBy(stats.cpu.min, cpuMin);
```
Comment (danielpeach) [inline app/scripts/modules/azure/instance/azureInstanceType.service.js]: Probably should be `Math.min` or `_.minBy`.

Comment (icfantv) [inline app/scripts/modules/azure/instance/azureInstanceType.service.js]: Ooo, yep.  Good catch.

File: app/scripts/modules/core/search/global/globalSearch.controller.js [hunk 2]
```diff
@@ -32,7 +32,7 @@ module.exports = angular.module('spinnaker.core.search.global.controller', [
 
     this.searchFieldBlurred = ($blurEvent) => {
       // if the target is outside the global search (e.g. shift+tab), hide the results
-      if (!$.contains($element.get(0), $blurEvent.relatedTarget)) {
+      if (!$.includes($element.get(0), $blurEvent.relatedTarget)) {
```
Comment (danielpeach) [inline app/scripts/modules/core/search/global/globalSearch.controller.js]: Is this right? I don't know enough about jQuery to know.

Comment (icfantv) [inline app/scripts/modules/core/search/global/globalSearch.controller.js]: No, that's incorrect as well.

File: app/scripts/modules/azure/instance/azureInstanceType.service.js (hunk 1 above)
Comment (icfantv) [inline app/scripts/modules/azure/instance/azureInstanceType.service.js]: Fixed.  I also verified the other `Math.m*` calls.

File: app/scripts/modules/core/search/global/globalSearch.controller.js (hunk 2 above)
Comment (icfantv) [inline app/scripts/modules/core/search/global/globalSearch.controller.js]: Fixed.

**Comment by icfantv** [comment]
The rebase pulled in code that wasn't right so the lodash module import and `_` DI had to be removed from `modules/google/serverGroup/configure/wizard/advancedSettings/advancedSettingsSelector.directive.js` as well.  This is fixed.

File: app/scripts/modules/google/serverGroup/configure/serverGroupConfiguration.service.js [hunk 3]
```diff
@@ -410,12 +412,12 @@ module.exports = angular.module('spinnaker.serverGroup.configure.gce.configurati
...
-      filteredData.subnets = _(command.backingData.subnets)
+      filteredData.subnets = _.chain(command.backingData.subnets)
         .filter({ account: command.credentials, network: command.network, region: command.region })
-        .pluck('name')
-        .valueOf();
+        .map('name')
+        .value();
 
-      if (!_(filteredData.subnets).contains(command.subnet)) {
+      if (!_.chain(filteredData.subnets).includes(command.subnet).value()) {
```
Comment (anotherchrisberry) [inline app/scripts/modules/google/serverGroup/configure/serverGroupConfiguration.service.js]: looks like this can just be `if (!filteredData.subnets.includes(command.subnet)) {`

File: app/scripts/modules/google/serverGroup/configure/serverGroupConfiguration.service.js [hunk 4]
```diff
@@ -439,23 +441,23 @@ module.exports = angular.module('spinnaker.serverGroup.configure.gce.configurati
       if (currentOptions && command.securityGroups) {
         // not initializing - we are actually changing groups
         var currentGroupNames = command.securityGroups.map(function(groupId) {
-          var match = _(currentOptions).find({id: groupId});
+          var match = _.chain(currentOptions).find({id: groupId}).value();
```
Comment (anotherchrisberry) [inline app/scripts/modules/google/serverGroup/configure/serverGroupConfiguration.service.js]: `var match = _.find(currentOptions, {id: groupId})`

File: app/scripts/modules/google/serverGroup/configure/serverGroupConfiguration.service.js [hunk 5]
```diff
@@ -439,23 +441,23 @@ module.exports = angular.module('spinnaker.serverGroup.configure.gce.configurati
...
           return match ? match.name : groupId;
         });
 
         var matchedGroups = command.securityGroups.map(function(groupId) {
-          var securityGroup = _(currentOptions).find({id: groupId}) ||
-              _(currentOptions).find({name: groupId});
+          var securityGroup = _.chain(currentOptions).find({id: groupId}).value() ||
+              _.chain(currentOptions).find({name: groupId}).value();
```
Comment (anotherchrisberry) [inline app/scripts/modules/google/serverGroup/configure/serverGroupConfiguration.service.js]: same as above (no need to `_.chain`)

File: app/scripts/modules/google/serverGroup/configure/serverGroupConfiguration.service.js [hunk 6]
```diff
@@ -439,23 +441,23 @@ module.exports = angular.module('spinnaker.serverGroup.configure.gce.configurati
...
           return securityGroup ? securityGroup.name : null;
         }).map(function(groupName) {
-          return _(newSecurityGroups).find({name: groupName});
+          return _.chain(newSecurityGroups).find({name: groupName}).value();
```
Comment (anotherchrisberry) [inline app/scripts/modules/google/serverGroup/configure/serverGroupConfiguration.service.js]: skip `chain` here, too

File: app/scripts/modules/google/serverGroup/configure/wizard/securityGroups/tagManager.service.js [hunk 7]
```diff
@@ -106,13 +106,13 @@ module.exports = angular.module('spinnaker.deck.gce.tagManager.service', [
         tags = c.tags,
         securityGroupObjectsWithTag = this.securityGroupObjectsKeyedByTag[tagName];
 
-      if (!_.contains(tags.map(t => t.value), tagName)) {
+      if (!_.includes(tags.map(t => t.value), tagName)) {
```
Comment (anotherchrisberry) [inline app/scripts/modules/google/serverGroup/configure/wizard/securityGroups/tagManager.service.js]: `if (tags.map(t => t.value).includes(tagName)) {`

File: app/scripts/modules/netflix/fastProperties/modal/fastPropertyUpsert.controller.js [hunk 8]
```diff
@@ -108,7 +109,7 @@ module.exports = angular
     vm.property.targetScope = angular.copy(fastProperty.selectedScope) || {};
 
     vm.strategies = fastPropertyStrategy.getStrategies();
-    vm.selectedStrategy = vm.strategies.length === 1 ? {selected: _.first(vm.strategies)} : {};
+    vm.selectedStrategy = vm.strategies.length === 1 ? {selected: _.head(vm.strategies)} : {};
```
Comment (anotherchrisberry) [inline app/scripts/modules/netflix/fastProperties/modal/fastPropertyUpsert.controller.js]: `vm.strategies[0]` would read a little easier here, since we already know there's exactly one item.

File: app/scripts/modules/netflix/instance/titus/netflixTitusInstanceDetails.controller.js [hunk 9]
```diff
@@ -47,7 +48,7 @@ module.exports = angular.module('spinnaker.netflix.instance.titus.controller', [
     this.getBastionAddressForAccount = (account, region) => {
       return accountService.getAccountDetails(account).then((details) => {
         this.bastionHost = details.bastionHost || 'unknown';
-        this.apiEndpoint = _.where(details.regions, {name: region})[0].endpoint;
+        this.apiEndpoint = _.filter(details.regions, {name: region})[0].endpoint;
```
Comment (anotherchrisberry) [inline app/scripts/modules/netflix/instance/titus/netflixTitusInstanceDetails.controller.js]: can use `_.find` here and then skip the `[0]`

**Comment by anotherchrisberry** [comment]
A few other `_.chain` calls that can be dropped:
netflix/serverGroup/wizard/serverGroupCommandConfigurer.service.js#L20
openstack/serverGroup/wizard/serverGroupConfiguration.service.js#L200

`_.filter(...)[0]` that can be `.find`:
titus/pipeline/stages/runJob/runJobExecutionDetails.controller.js#L28
titus/serverGroup/details/serverGroupDetails.controller.js#L50

File: app/scripts/modules/netflix/instance/titus/netflixTitusInstanceDetails.controller.js (hunk 9 above)
Comment (icfantv) [inline app/scripts/modules/netflix/instance/titus/netflixTitusInstanceDetails.controller.js]: @anotherchrisberry Yea, there's a bunch of places where we can drop lodash and use es6 API functions but I thought it better to do that in a separate PR.  Thoughts?

**Comment by anotherchrisberry** [comment]
@spinnaker/microsoft can someone pull this branch down and see if anything is broken on the Azure side?
same applies to @gregturn for the Cloud Found module, @gadresushrut or @drmaas or someone from the Openstack team.

We've reviewed the actual changes, but should run the application and watch for any errors in the console to see if we've missed anything that was not changed.

File: app/scripts/modules/amazon/serverGroup/details/scalingPolicy/chart/metricAlarmChart.component.spec.js [hunk 10]
```diff
@@ -4,7 +4,7 @@ import {Subject} from 'rxjs';
 
 describe('Component: metricAlarmChart', function () {
 
-  var $ctrl, $scope, cloudMetricsReader, $q;
+  var $ctrl, $scope, cloudMetricsReader, rx, $q;
```
Comment (icfantv) [inline app/scripts/modules/amazon/serverGroup/details/scalingPolicy/chart/metricAlarmChart.component.spec.js]: crud.  i bet this was from the rebase....this is going to stink...

Comment (icfantv) [inline app/scripts/modules/amazon/serverGroup/details/scalingPolicy/chart/metricAlarmChart.component.spec.js]: ok, i think this was just an errant non-deletion.  this was the only case.  fixed.

File: app/scripts/modules/azure/instance/azureInstanceType.service.js [hunk 11]
```diff
@@ -225,23 +226,23 @@ module.exports = angular.module('spinnaker.azure.instanceType.service', [
...
+          stats.cpu.max = Math.maxBy(stats.cpu.max, cpuMax);
+          stats.memory.min = Math.minBy(stats.memory.min, memoryMin);
+          stats.memory.max = Math.maxBy(stats.memory.max, memoryMax);
+          stats.storage.min = Math.minBy(stats.storage.min, storageMin);
+          stats.storage.max = Math.maxBy(stats.storage.max, storageMax);
+          stats.costFactor.min = Math.minBy(stats.costFactor.min, costFactorMin);
+          stats.costFactor.max = Math.maxBy(stats.costFactor.max, costFactorMax);
```
Comment (icfantv) [inline app/scripts/modules/azure/instance/azureInstanceType.service.js]: Yea.  These are fixed.

File: app/scripts/modules/netflix/instance/titus/netflixTitusInstanceDetails.controller.js (hunk 9 above)
Comment (anotherchrisberry) [inline app/scripts/modules/netflix/instance/titus/netflixTitusInstanceDetails.controller.js]: Agreed - as long as everything is _functional_ in this PR, we can go ahead and merge it.

**Comment by anotherchrisberry** [comment]
@danielpeach @lwander have you had a chance to pull this down and kick the tires?

**Comment by danielpeach** [comment]
I tested for kubernetes and GCE -- no issues. Thanks for doing this.

**Comment by anotherchrisberry** [comment]
@icfantv I think it's safe to rebase and merge this when you have a chance.

**Comment by anotherchrisberry** [comment]
@icfantv Just talked to @scotmoor - he's going to test the Azure side of things and let you know when he's good with it. Let's hold off on merging until then.

**Comment by icfantv** [comment]
@anotherchrisberry 👍

**Comment by scotmoor** [comment]
done testing Azure. No issues here

**Comment by gregturn** [comment]
@anotherchrisberry I tested against CF and don't see any issues.

**Comment by anotherchrisberry** [comment]
Sounds like we're good to rebase and merge.

**Comment by icfantv** [comment]
Woot!  I'll do this tonight.
